LPC SoC family GPIO interrupts

agansari opened the pull request:
LPC54xxx and LPC55xxx GPIO interrupts enables.
PINT and INPUTMUX devices enabled to route interrupt signal from pin to NVIC.

Tested with GPIO driver sample.

Fixes #16167

agansari committed:
boards: lpcxpresso55s69 pinmux macros refactor

Board is refactored to use DTS generated value, not use
magic numbers.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
drivers: gpio_mcux_lpc GPIO interrupts

LPC GPIO architecture uses multiple devices.
GPIO input is routed via INPUTMUX to the PINT
device which roots the interrupt to NVIC.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
dts: NXP LPC gpio updates

NXP's LPC family of MCU's GPIOs parameters is udated.
Boards LPC54xxx and LPC55xxx have updated values according
pin and interrupt layout.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

MaureenHelm commented on ext/hal/nxp/mcux/drivers/lpc/CMakeLists.txt:
```diff
@@ -6,4 +6,5 @@
 
-zephyr_sources_ifdef(CONFIG_USART_MCUX_LPC    fsl_usart.c fsl_flexcomm.c)
+zephyr_sources_ifdef(CONFIG_USART_MCUX_LPC  fsl_usart.c fsl_flexcomm.c)
+zephyr_sources_ifdef(CONFIG_GPIO_MCUX_LPC   fsl_gpio.c fsl_inputmux.c)
```
This change is in the wrong commit

MaureenHelm commented on dts/bindings/gpio/nxp,lpc-gpio.yaml:
```diff
@@ -0,0 +1,41 @@
+      category: required
+      description: compatible strings
+      constraint: "nxp,lpc-gpio"
+      generation: define
```
Include base.yaml and remove the properties that are now defined there.

agansari replied:
@MaureenHelm I revisited this pull yesterday, and realized after a rebase that I need to do some rework on this pull, including this .yaml.

agansari replied:
@MaureenHelm i've reworked this pull, please review it again.

MaureenHelm commented on boards/arm/lpcxpresso55s69/pinmux.c:
```diff
@@ -84,8 +86,12 @@ static int lpcxpresso_55s69_pinmux_init(struct device *dev)
-	pinmux_pin_set(port1, 9, port1_pin9_config);
+#if defined(CONFIG_GPIO_MCUX_LPC_PORT0) || defined(CONFIG_GPIO_MCUX_LPC_PORT1)
+	/* Turn on PINT device*/
+	PINT_Init(PINT);
```
Would it be better to do this in soc.c instead?

MaureenHelm commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -109,49 +144,226 @@ static int gpio_mcux_lpc_read(struct device *dev,
+	INPUTMUX_AttachSignal(
+		INPUTMUX, pint_idx,
+		PIN_TO_INPUT_MUX_CONNECTION(port, DT_ALIAS_SW0_GPIOS_PIN));
+	data->pint_id[DT_ALIAS_SW0_GPIOS_PIN] = pint_idx++;
```
Using `DT_ALIAS_SW0_GPIOS_PIN` here means that gpio interrupts will only work for the aliased buttons/switches. It will not work in the general case.

agansari replied:
@MaureenHelm What do you mean by "It will not work in the general case."?
InputMux device will map one pin to one interrupt (manual 19.6.3/4).

Analyzing the way I allocate PINT association, I'm not happy with the way I did it.
I'd rather have a static dts allocation, or even better, deduce PINT allocation according to isr no.

MaureenHelm replied:
You're always using the SW0 pin. What if I want to use a sensor pin for data ready interrupts?

agansari replied:
@MaureenHelm okay, I will change SW to something generic

agansari replied:
@MaureenHelm made the changes required, please review it again.
Thank you for your reviews!

agansari committed:
soc: enable PINT device at LPC SoC boot

PINT device is enabled when SoC is booting up. Applies to LPC54xxx and
LPC55xxx families.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
west.yaml: Update GPIO related hal_nxp

Get separate LPC pint devices and GPIO includes.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
@MaureenHelm I rebased and done the west split for this issue, please review.

galak commented:
We should have this target the new GPIO APIs and such.  I'd say lets pend on this til the initial commit is in the `topic-gpio` branch and we can target that there.

However it probably makes sense to review https://github.com/zephyrproject-rtos/hal_nxp/pull/11 and have that merged ASAP.

agansari commented:
Closed via #19714, feature will be available once #18530 is merged
